### PR TITLE
Release/v2.6.0

### PR DESCRIPTION
**Pull request by tailingchen**
Release v2.6.0 is based on release v2.5.0, and includes the following features:
- https://github.com/perpetual-protocol/perp-curie-contract/pull/36

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,24 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- Add `AccountBalance.getMarketRegistry()` to return the address of MarketRegistry contract.
+- Add `AccountBalance.getMarkPrice(address)` to return the mark price of given market.
+- Add `ClearingHouseConfig.getMarkPriceConfig()` to return marketTwapInterval and premiumInterval used for mark price calculations.
+- Add new event `MarketRegistryChanged` to `AccountBalance`, will emit in `setMarketRegistry(address)`
+
+### Changed
+- Move events in ClearingHouseConfig to IClearingHouseConfigEvent in IClearingHouseConfig.sol
+
+### Deprecated
+- BackstopLiquidityProvider from ClearingHouseConfig & IClearingHouseConfig and comments added to ClearingHouseConfigStorage
```
Comment (vinta) [inline CHANGELOG.md:25]: What is the reason we deprecated `backstopLiquidityProvider`? Any doc or discussion as references?

File: contracts/BaseToken.sol [hunk 2]
```diff
@@ -32,7 +32,7 @@ contract BaseToken is IBaseToken, IIndexPrice, VirtualToken, BlockContext, BaseT
     ) external initializer {
         __VirtualToken_init(nameArg, symbolArg);
 
-        uint8 priceFeedDecimals = IPriceFeedV2(priceFeedArg).decimals();
+        uint8 priceFeedDecimals = IPriceFeedDispatcher(priceFeedArg).decimals();
```
Comment (vinta) [inline contracts/BaseToken.sol:35]: 1. When deploying or upgrading a baseToken, it will have one `PriceFeedDispatcher` contract, one `ChainlinkPriceFeedV3` contract, and one `UniswapV3PriceFeed` contract?

2. So `baseToken.getPriceFeed()` returns the address of `PriceFeedDispatcher` instead of the current underlying price feed (ChainlinkPriceFeedV3 or UniswapV3PriceFeed)? 
Should we rename it to `getPriceFeedDispatcher()` since the behaviour changes?
  2-1. Maybe have both `getPriceFeed()` and `getPriceFeedDispatcher()` for backward compatibility.

3. If 2 is the case, we will need to modify `LimitOrderBook`. Currently, it is using `getPriceFeed()` and assume it is a priceFeed and has `getRoundData(roundId)`.
Added a TODO: https://github.com/perpetual-protocol/perp-curie-periphery-contract/pull/35

4. When pausing a market, we don't use `EmergencyPriceFeed` anymore right?

Comment (tailingchen) [reply contracts/BaseToken.sol:35]: > 1. When deploying or upgrading a baseToken, it will have one PriceFeedDispatcher contract, one ChainlinkPriceFeedV3 contract, and one UniswapV3PriceFeed contract?

Yes

> 4. When pausing a market, we don't use EmergencyPriceFeed anymore right?

Yes

@twxia @tienshaoku  could you help to answer Q2 and Q3?

File: CHANGELOG.md (hunk 1 above)
Comment (tailingchen) [reply CHANGELOG.md:25]: The `backstopLiquidityProvider` is whitelist liquidators to liquidate the account has bad debt. It's the short-term solution to prevent the bad debt attack.

Now we already released new liquidation mechanism, the long-term solution of bad debt attack. Current liquidation is through taking over positions instead of market selling position.

Check the [internal docs](https://www.notion.so/perp/3799-4134-Bug-Reports-276fe67b8cb44578a7eda7bace70f77a#9b573a42c01f45b28289c05332132660 ) for details.

File: contracts/BaseToken.sol (hunk 2 above)
Comment (tienshaoku) [reply contracts/BaseToken.sol:35]: @vinta 2 yes i'll discuss it with William!
3 thx for pointing that out! will do!

File: contracts/AccountBalance.sol [hunk 3]
```diff
@@ -332,6 +334,11 @@ contract AccountBalance is IAccountBalance, BlockContext, ClearingHouseCallee, A
         return positionSize.mulRatio(maxLiquidateRatio);
     }
 
+    /// @inheritdoc IAccountBalance
+    function getMarkPrice(address baseToken) external view override returns (uint256) {
+        return _getMarkPrice(baseToken);
+    }
```
Comment (vinta) [inline contracts/AccountBalance.sol:338]: Not a blocker: could we have a section in [Mark Price Spec](https://www.notion.so/perp/Mark-Price-Spec-96c9651076f5485d8e38e2fe64bc9f76) that indicates which components are using markPrice, and which aren't?

From my understanding:

- positionValue => markPrice
- accountValue => markPrice
- freeCollateral => markPrice
- marginRequirement => markPrice
- unrealizedPnl => markPrice
- liquidation => markPrice
- isOverPriceSpread => spread between marketPrice and indexPrice 
  - Though `Exchange.getSqrtMarkTwapX96(baseToken, 0)`is actually marketPrice instead of markPrice
  - Might want to rename?
- funding => spread between marketPrice (TWAP) and indexPrice (TWAP)
  - Similar to the above case
- Not sure this list is complete or not

File: contracts/ClearingHouse.sol [hunk 4]
```diff
@@ -14,24 +14,23 @@ import { PerpSafeCast } from "./lib/PerpSafeCast.sol";
...
 import { Funding } from "./lib/Funding.sol";
+import { AccountMarket } from "./lib/AccountMarket.sol";
+import { OpenOrder } from "./lib/OpenOrder.sol";
 import { OwnerPausable } from "./base/OwnerPausable.sol";
+import { BlockContext } from "./base/BlockContext.sol";
 import { IERC20Metadata } from "./interface/IERC20Metadata.sol";
 import { IVault } from "./interface/IVault.sol";
 import { IExchange } from "./interface/IExchange.sol";
 import { IOrderBook } from "./interface/IOrderBook.sol";
 import { IIndexPrice } from "./interface/IIndexPrice.sol";
```
Comment (vinta) [inline contracts/ClearingHouse.sol]: Can remove `IIndexPrice` which is unused now.

**Review by vinta — DISMISSED** [review]

File: contracts/BaseToken.sol (hunk 2 above)
Comment (twxia) [reply contracts/BaseToken.sol:35]: for 3., 

I have replied in https://github.com/perpetual-protocol/perp-curie-periphery-contract/pull/35/files#r1053484472.

TLDR, limit order needs historical data which is not supported in v3. probabbly remain the same is fine.

File: contracts/AccountBalance.sol (hunk 3 above)
Comment (tailingchen) [reply contracts/AccountBalance.sol:338]: Already have [Impact Scope ](https://www.notion.so/perp/Mark-Price-Spec-96c9651076f5485d8e38e2fe64bc9f76#71da12cedbc64fe3949ab8b9bc3f980a)session in spec. Will list the non-impact part in the doc.

File: contracts/BaseToken.sol (hunk 2 above)
Comment (vinta) [reply contracts/BaseToken.sol:35]: @tienshaoku @twxia Please check the reply https://github.com/perpetual-protocol/perp-curie-periphery-contract/pull/35/files#r1053936603 when you have time.

File: contracts/ClearingHouse.sol (hunk 4 above)
Comment (tailingchen) [reply contracts/ClearingHouse.sol]: Fixed in https://github.com/perpetual-protocol/perp-curie-contract/pull/54

File: contracts/AccountBalance.sol (hunk 3 above)
Comment (tailingchen) [reply contracts/AccountBalance.sol:338]: Rename markTwap to marketTwap in https://github.com/perpetual-protocol/perp-curie-contract/pull/54

File: contracts/Exchange.sol [hunk 5]
```diff
@@ -247,11 +247,17 @@ contract Exchange is
...
 
-        // if updating TWAP fails, this call will be reverted and thus using try-catch
-        try IBaseToken(baseToken).cacheTwap(IClearingHouseConfig(_clearingHouseConfig).getTwapInterval()) {} catch {}
-        uint256 markTwap;
+        // The purpose of caching index twap here is to save the gas consumption of calculating mark price,
+        // if updating TWAP fails, this call will be reverted and thus using try-catch.
+        // NOTE: the cached index twap is used for AccountBalance.MarkPrice calculation,
+        // not for funding rate calculation.
+        (, uint32 premiumInterval) = IClearingHouseConfig(_clearingHouseConfig).getMarkPriceConfig();
+        try IBaseToken(baseToken).cacheTwap(premiumInterval) {} catch {}
```
Comment (Wraecca) [inline contracts/Exchange.sol:237]: if L254-255 is not relevant to `settleFunding` suggest to move to other function, but I know it's limited by ClearingHouse bytecode size so let's worry about this in next few release

File: contracts/Exchange.sol [hunk 6]
```diff
@@ -247,11 +247,17 @@ contract Exchange is
...
+
+        // markTwap here is market twap, and used for funding rate calculation.
+        uint256 marketTwap;
```
Comment (Wraecca) [inline contracts/Exchange.sol:239]: instead of writing a comment suggest renaming directly to sth like marketTwapForFunding

File: contracts/storage/BaseTokenStorage.sol [hunk 7]
```diff
@@ -9,10 +9,12 @@ import { IBaseToken } from "../interface/IBaseToken.sol";
 abstract contract BaseTokenStorageV1 {
     // --------- IMMUTABLE ---------
 
+    // _priceFeedDecimals is now priceFeedDispatcherDecimals, which is IPriceFeedDispatcher.decimals()
     uint8 internal _priceFeedDecimals;
```
Comment (Wraecca) [inline contracts/storage/BaseTokenStorage.sol:13]: shall we just rename it to _priceFeedDispatcherDecimals if it won't break the upgrade?

File: contracts/AccountBalance.sol [hunk 8]
```diff
@@ -512,6 +517,43 @@ contract AccountBalance is IAccountBalance, BlockContext, ClearingHouseCallee, A
...
+            );
+        uint256 indexWithPremium = _getIndexPrice(baseToken, 0).toInt256().add(premium).toUint256();
+        return PerpMath.findMedianOfThree(marketPrice, marketTwap, indexWithPremium);
+    }
+
+    function _getMarketPrice(address baseToken, uint32 twapInterval) internal view returns (uint256) {
+        address marketRegistry = UniswapV3CallbackBridge(_orderBook).getMarketRegistry();
+        address pool = IMarketRegistry(marketRegistry).getPool(baseToken);
+        return
+            UniswapV3Broker.getSqrtMarketTwapX96(pool, twapInterval).formatSqrtPriceX96ToPriceX96().formatX96ToX10_18();
```
Comment (Wraecca) [inline contracts/AccountBalance.sol]: sorry i was wrong, i think it's better to not access uniswap directly if possible. we used to only use exchange and orderbook (for range order) to access uniswap (except market registry), and it'd be better if we keep the abstraction level the same as before

in this case i'd suggest to use IOrderBook(_orderbook).geExchange then query the getSqrtMarketTwapX96

File: contracts/AccountBalance.sol [hunk 9]
```diff
@@ -512,6 +517,43 @@ contract AccountBalance is IAccountBalance, BlockContext, ClearingHouseCallee, A
...
+    }
+
+    function _getIndexPrice(address baseToken, uint32 twapInterval) internal view returns (uint256) {
+        return IIndexPrice(baseToken).getIndexPrice(twapInterval);
+    }
+
+    function _isMarkPriceEnabled(uint32 marketTwapInterval, uint32 premiumInterval) internal view returns (bool) {
```
Comment (Wraecca) [inline contracts/AccountBalance.sol]: can be pure (internal auditor please remember to run any static analyzer like slither!)

File: contracts/Exchange.sol (hunk 6 above)
Comment (tailingchen) [reply contracts/Exchange.sol:239]: My bad, I should remove the comment after I renamed the variable from markTwap to marketTwap.

The name of marketTwap is correct.

File: contracts/AccountBalance.sol (hunk 9 above)
Comment (tailingchen) [reply contracts/AccountBalance.sol]: Fixed in https://github.com/perpetual-protocol/perp-curie-contract/pull/55

File: contracts/AccountBalance.sol (hunk 8 above)
Comment (tailingchen) [reply contracts/AccountBalance.sol]: agree, it's better to keep abstraction layer.
Fixed in https://github.com/perpetual-protocol/perp-curie-contract/pull/55

File: contracts/storage/BaseTokenStorage.sol (hunk 7 above)
Comment (tailingchen) [reply contracts/storage/BaseTokenStorage.sol:13]: It's fine for me. Even though it gets decimal from PriceFeedDispatcher, the decimal still represents the decimal of a price feed, not a dispatcher. the name of `_priceFeedDecimals ` is more reasonable than `_priceFeedDispatcherDecimals`.

@tienshaoku  what do you think?

File: contracts/Exchange.sol (hunk 6 above)
Comment (tailingchen) [reply contracts/Exchange.sol:239]: Fixed in https://github.com/perpetual-protocol/perp-curie-contract/pull/56

**Review by paco0x — DISMISSED** [review]

**Review by 42bchen — APPROVED** [review]
